java/vector/src/main/java/org/apache/arrow/vector/BaseLargeVariableWidthVector.java

```diff
@@ -0,0 +1,1361 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.vector;
+
+import static org.apache.arrow.memory.util.LargeMemoryUtil.capAtMaxInt;
+
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+import org.apache.arrow.memory.BaseAllocator;
+import org.apache.arrow.memory.BufferAllocator;
+import org.apache.arrow.memory.OutOfMemoryException;
+import org.apache.arrow.memory.util.ArrowBufPointer;
+import org.apache.arrow.memory.util.ByteFunctionHelpers;
+import org.apache.arrow.memory.util.hash.ArrowBufHasher;
+import org.apache.arrow.util.Preconditions;
+import org.apache.arrow.vector.compare.VectorVisitor;
+import org.apache.arrow.vector.ipc.message.ArrowFieldNode;
+import org.apache.arrow.vector.types.pojo.Field;
+import org.apache.arrow.vector.util.CallBack;
+import org.apache.arrow.vector.util.OversizedAllocationException;
+import org.apache.arrow.vector.util.TransferPair;
+
+import io.netty.buffer.ArrowBuf;
+
+/**
+ * BaseLargeVariableWidthVector is a base class providing functionality for large strings/large bytes types.
+ */
+public abstract class BaseLargeVariableWidthVector extends BaseValueVector
```

Comment:
Does it make sense for this to extend BaseVariableWidthVector?

Reply:
We might be able to reuse quite some functionality. However, generally in Arrow we have been okay with some code duplication to avoid the performance overhead of having a deep inheritance hierarchy. So may be the current approach is fine. Just wondering if we evaluated this.

Reply:
Good question.
We choose to have a separate BaseLargeVariableWidthVector class after careful thoughts and evaluations.

It appears that BaseVariableWidthVector and BaseLargeVariableWidthVector have many similarities. However, they are different in details, and the differences cannot be overcome by method overriding/overloading.

For example, in BaseVariableWidthVector, we have

protected final int getStartOffset(int index)

while in BaseLargeVariableWidthVector, we have

protected final long getStartOffset(int index)

These two methods differ only in return type, so we cannot use method overloading/overriding.

Another concern is performance. For example, `handleSafe` is a performance critical operation, because it may be called in each setSafe method. So it is declared as a final method. If we share the same base class, we must provide two implementations for `handleSafe`, and it will no longer be fianl. This may lead to performance degradation.